File: mac-athlon-agent/AthlonAgent/Core/SessionTurnReconciler.swift
import Foundation

struct SessionTurnEndSnapshot {
    var assistantContent: String?
    var assistantReasoning: String?
    var incompleteToolCalls: [AgentToolCall]
    var wasCancelled: Bool
    var timedOut: Bool
    var errorMessage: String?
}

struct SessionTurnReconcileResult {
    let session: AgentSession
    let persistedMessages: [ChatMessage]
}

enum SessionTurnReconciler {
    static func reconcile(_ session: AgentSession, snapshot: SessionTurnEndSnapshot) -> SessionTurnReconcileResult {
        guard needsReconcile(snapshot) else {
            return SessionTurnReconcileResult(session: session, persistedMessages: [])
        }

        var persisted: [ChatMessage] = []
        var messages = session.messages
        let parentId = findLastUserMessageId(messages)
        var answered = buildAnsweredToolCallIds(messages)

        let incompleteTools = snapshot.incompleteToolCalls
            .filter { !$0.id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .reduce(into: [AgentToolCall]()) { acc, call in
                if !acc.contains(where: { $0.id == call.id }) { acc.append(call) }
            }
            .filter { !answered.contains($0.id) }

        let hasAssistantText = !(snapshot.assistantContent ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            || !(snapshot.assistantReasoning ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        let tailAssistant = findTailAssistantAfterLastUser(messages)

        if tailAssistant == nil, hasAssistantText || !incompleteTools.isEmpty {
            var content = snapshot.assistantContent ?? ""
            if content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, snapshot.wasCancelled {
                content = "（生成已停止）"
            }
            let assistant = ChatMessage(
                role: .assistant,
                content: content,
                reasoningContent: snapshot.assistantReasoning ?? "",
                toolCalls: incompleteTools.isEmpty ? nil : incompleteTools,
                parentMessageId: parentId
            )
            messages.append(assistant)
            persisted.append(assistant)
            for call in incompleteTools { answered.insert(call.id) }
        }

        for toolCall in incompleteTools where !answered.contains(toolCall.id) {
            let toolMessage = ChatMessage(
                role: .tool,
                content: AgentRuntimeToolFormatting.formatToolResult(
                    toolCall,
                    buildInterruptedToolResult(snapshot)
                ),
                parentMessageId: parentId
            )
            messages.append(toolMessage)
            persisted.append(toolMessage)
            answered.insert(toolCall.id)
        }

        if let notice = buildTurnNotice(snapshot), !notice.isEmpty {
            let systemMessage = ChatMessage(role: .system, content: notice, parentMessageId: parentId)
            messages.append(systemMessage)
            persisted.append(systemMessage)
        }

        guard !persisted.isEmpty else {
            return SessionTurnReconcileResult(session: session, persistedMessages: [])
        }
        return SessionTurnReconcileResult(session: session.withMessages(messages), persistedMessages: persisted)
    }

    private static func needsReconcile(_ snapshot: SessionTurnEndSnapshot) -> Bool {
        snapshot.wasCancelled || snapshot.timedOut || !(snapshot.errorMessage ?? "").isEmpty
    }

    private static func buildInterruptedToolResult(_ snapshot: SessionTurnEndSnapshot) -> ToolResult {
        if let error = snapshot.errorMessage, !error.isEmpty {
            return .failure(summary: "工具未完成", error: "模型调用失败，工具未执行完成：\(error)")
        }
        if snapshot.timedOut {
            return .failure(summary: "工具未完成", error: "本回合因超时被自动停止，工具未执行完成。")
        }
        return .failure(
            summary: "工具未完成",
            error: "上次对话在工具执行或生成时被用户停止。可继续发送消息让助手接着处理。"
        )
    }

    private static func buildTurnNotice(_ snapshot: SessionTurnEndSnapshot) -> String? {
        if let error = snapshot.errorMessage, !error.isEmpty { return error }
        if snapshot.timedOut { return "本回合已超过配置的超时时间，已自动停止。" }
        if snapshot.wasCancelled { return "生成已停止。" }
        return nil
    }

    private static func findLastUserMessageId(_ messages: [ChatMessage]) -> String? {
        messages.last(where: { $0.role == .user })?.id
    }

    private static func findTailAssistantAfterLastUser(_ messages: [ChatMessage]) -> ChatMessage? {
        guard let lastUserIndex = messages.lastIndex(where: { $0.role == .user }) else { return nil }
        return messages[(lastUserIndex + 1)...].first(where: { $0.role == .assistant })
    }

    private static func buildAnsweredToolCallIds(_ messages: [ChatMessage]) -> Set<String> {
        var answered = Set<String>()
        for message in messages where message.role == .tool {
            if let id = extractToolCallId(message.content) {
                answered.insert(id)
            }
        }
        return answered
    }

    private static func extractToolCallId(_ content: String?) -> String? {
        guard let content else { return nil }
        for line in content.split(whereSeparator: \.isNewline) {
            let text = String(line)
            if text.lowercased().hasPrefix("toolcallid:") {
                let value = text.dropFirst("toolcallid:".count).trimmingCharacters(in: .whitespaces)
                return value.isEmpty ? nil : value
            }
        }
        return nil
    }
}

File: athlon-work/src/Athlon.Agent.Core/SessionTurnReconciler.cs
namespace Athlon.Agent.Core;

public sealed record SessionTurnEndSnapshot(
    string? AssistantContent,
    string? AssistantReasoning,
    IReadOnlyList<AgentToolCall> IncompleteToolCalls,
    bool WasCancelled,
    bool TimedOut,
    string? ErrorMessage);

public sealed record SessionTurnReconcileResult(AgentSession Session, IReadOnlyList<ChatMessage> PersistedMessages);

public static class SessionTurnReconciler
{
    public static SessionTurnReconcileResult Reconcile(AgentSession session, SessionTurnEndSnapshot snapshot)
    {
        if (!NeedsReconcile(snapshot))
        {
            return new SessionTurnReconcileResult(session, Array.Empty<ChatMessage>());
        }

        var persisted = new List<ChatMessage>();
        var messages = session.Messages.ToList();
        var parentId = FindLastUserMessageId(messages);
        var answeredToolCallIds = BuildAnsweredToolCallIds(messages);

        var incompleteTools = snapshot.IncompleteToolCalls
            .Where(call => !string.IsNullOrWhiteSpace(call.Id))
            .Where(call => !answeredToolCallIds.Contains(call.Id))
            .GroupBy(call => call.Id, StringComparer.Ordinal)
            .Select(group => group.First())
            .ToList();

        var hasAssistantText = !string.IsNullOrWhiteSpace(snapshot.AssistantContent)
            || !string.IsNullOrWhiteSpace(snapshot.AssistantReasoning);
        var tailAssistant = FindTailAssistantAfterLastUser(messages);

        if (tailAssistant is null && (hasAssistantText || incompleteTools.Count > 0))
        {
            var content = snapshot.AssistantContent ?? string.Empty;
            if (string.IsNullOrWhiteSpace(content) && snapshot.WasCancelled)
            {
                content = "（生成已停止）";
            }

            var assistant = ChatMessage.Create(
                MessageRole.Assistant,
                content,
                parentId,
                incompleteTools.Count > 0 ? incompleteTools : null,
                snapshot.AssistantReasoning);
            messages.Add(assistant);
            persisted.Add(assistant);

            foreach (var toolCall in incompleteTools)
            {
                answeredToolCallIds.Add(toolCall.Id);
            }
        }

        foreach (var toolCall in incompleteTools)
        {
            if (answeredToolCallIds.Contains(toolCall.Id))
            {
                continue;
            }

            var toolMessage = ChatMessage.Create(
                MessageRole.Tool,
                ToolInvoker.FormatToolResult(toolCall, BuildInterruptedToolResult(snapshot)),
                parentId);
            messages.Add(toolMessage);
            persisted.Add(toolMessage);
            answeredToolCallIds.Add(toolCall.Id);
        }

        var notice = BuildTurnNotice(snapshot);
        if (!string.IsNullOrWhiteSpace(notice))
        {
            var systemMessage = ChatMessage.Create(MessageRole.System, notice, parentId);
            messages.Add(systemMessage);
            persisted.Add(systemMessage);
        }

        if (persisted.Count == 0)
        {
            return new SessionTurnReconcileResult(session, Array.Empty<ChatMessage>());
        }

        return new SessionTurnReconcileResult(session.WithMessages(messages), persisted);
    }

    private static bool NeedsReconcile(SessionTurnEndSnapshot snapshot) =>
        snapshot.WasCancelled
        || snapshot.TimedOut
        || !string.IsNullOrWhiteSpace(snapshot.ErrorMessage);

    private static ToolResult BuildInterruptedToolResult(SessionTurnEndSnapshot snapshot)
    {
        if (!string.IsNullOrWhiteSpace(snapshot.ErrorMessage))
        {
            return ToolResult.Failure(
                "工具未完成",
                $"模型调用失败，工具未执行完成：{snapshot.ErrorMessage}");
        }

        if (snapshot.TimedOut)
        {
            return ToolResult.Failure(
                "工具未完成",
                "本回合因超时被自动停止，工具未执行完成。");
        }

        return ToolResult.Failure(
            "工具未完成",
            "上次对话在工具执行或生成时被用户停止。可继续发送消息让助手接着处理。");
    }

    private static string? BuildTurnNotice(SessionTurnEndSnapshot snapshot)
    {
        if (!string.IsNullOrWhiteSpace(snapshot.ErrorMessage))
        {
            return snapshot.ErrorMessage;
        }

        if (snapshot.TimedOut)
        {
            return "本回合已超过配置的超时时间，已自动停止。";
        }

        if (snapshot.WasCancelled)
        {
            return "生成已停止。";
        }

        return null;
    }

    private static string? FindLastUserMessageId(IReadOnlyList<ChatMessage> messages)
    {
        for (var index = messages.Count - 1; index >= 0; index--)
        {
            if (messages[index].Role == MessageRole.User)
            {
                return messages[index].Id;
            }
        }

        return null;
    }

    private static ChatMessage? FindTailAssistantAfterLastUser(IReadOnlyList<ChatMessage> messages)
    {
        var lastUserIndex = -1;
        for (var index = messages.Count - 1; index >= 0; index--)
        {
            if (messages[index].Role == MessageRole.User)
            {
                lastUserIndex = index;
                break;
            }
        }

        if (lastUserIndex < 0)
        {
            return null;
        }

        for (var index = lastUserIndex + 1; index < messages.Count; index++)
        {
            if (messages[index].Role == MessageRole.Assistant)
            {
                return messages[index];
            }
        }

        return null;
    }

    private static HashSet<string> BuildAnsweredToolCallIds(IReadOnlyList<ChatMessage> messages)
    {
        var answered = new HashSet<string>(StringComparer.Ordinal);
        foreach (var message in messages)
        {
            if (message.Role != MessageRole.Tool)
            {
                continue;
            }

            var toolCallId = ExtractToolCallId(message.Content);
            if (!string.IsNullOrWhiteSpace(toolCallId))
            {
                answered.Add(toolCallId);
            }
        }

        return answered;
    }

    private static string? ExtractToolCallId(string? content) =>
        ToolInvoker.ExtractToolCallId(content);
}
